data_structures/linked_list/doubly_linked_list.py

```diff
@@ -9,76 +9,109 @@
-        # oldHead --> 2ndElement(head) nothing pointing at it so the old head will be
-        # removed
-        self.head.previous = None
-        if self.head is None:
-            self.tail = None  # if empty linked list
-        return temp
-
-    def insertTail(self, x):
-        newLink = Link(x)
-        newLink.next = None  # currentTail(tail)    newLink -->
-        self.tail.next = newLink  # currentTail(tail) --> newLink -->
-        newLink.previous = self.tail  # currentTail(tail) <--> newLink -->
-        self.tail = newLink  # oldTail <--> newLink(tail) -->
-
-    def deleteTail(self):
-        temp = self.tail
-        self.tail = self.tail.previous  # 2ndLast(tail) <--> oldTail --> None
-        self.tail.next = None  # 2ndlast(tail) --> None
-        return temp
-
-    def delete(self, x):
+            self.head.previous = new_node
+            new_node.next = self.head
+            self.head = new_node
+
+    def delete_head(self) -> str:
+        if self.is_empty:
+            return "List is empty"
+
+        head_data = self.head.data
+        if self.head.next:
+            self.head = self.head.next
+            self.head.previous = None
+
+        else:  # If there is no next previous node
+            self.head = None
+            self.tail = None
+
+        return head_data
+
+    def insert_at_tail(self, data):
+        new_node = Node(data)
+        self.tail.next = new_node
+        new_node.previous = self.tail
+        self.tail = new_node
+
+    def delete_tail(self) -> str:
+        if self.is_empty:
+            return "List is empty"
+
+        tail_data = self.tail.data
+        if self.tail.previous:
+            self.tail = self.tail.previous
+            self.tail.next = None
+        else:  # if there is no previous node
+            self.head = None
+            self.tail = None
+
+        return tail_data
+
+    def delete(self, data) -> str:
         current = self.head
 
-        while current.value != x:  # Find the position to delete
-            current = current.next
+        while current.data != data:  # Find the position to delete
+            if current.next:
+                current = current.next
+            else:  # We have reached the end an no value matches
+                return "No data matching given value"
 
         if current == self.head:
-            self.deleteHead()
+            self.delete_head()
 
         elif current == self.tail:
-            self.deleteTail()
+            self.delete_tail()
 
         else:  # Before: 1 <--> 2(current) <--> 3
             current.previous.next = current.next  # 1 --> 3
             current.next.previous = current.previous  # 1 <--> 3
+        return data
 
-    def isEmpty(self):  # Will return True if the list is empty
+    @property
+    def is_empty(self):  # Will return True if the list is empty
         return self.head is None
 
     def display(self):  # Prints contents of the list
```

Comment:
Could we please define `__str__()` instead and let the caller print or log or whatever.